### PR TITLE
Add vault variable example

### PR DESCRIPTION
**Pull request by tpmccallum**
Adding vault variable example, so that we can link to it from the documentation Re: [this developer documentation issue](https://github.com/fermyon/developer/pull/1157)

This example is meant to go hand-in-hand with the https://developer.fermyon.com/spin/v2/dynamic-configuration#vault-application-variable-provider-example

### CODE REVIEW

File: examples/vault-variable-test/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+description = "A simple application that returns query values from variable providers"
+version = "0.1.0"
+edition = "2021"
+
+[lib]
+crate-type = ["cdylib"]
+
+[dependencies]
+anyhow = "1"
+spin-sdk = { git = "https://github.com/fermyon/spin", tag = "v2.1.0" }
```
Comment (itowlson) [inline examples/vault-variable-test/Cargo.toml:13]:
```suggestion
spin-sdk = { path = "../../sdk/rust" }
```

We don't want to have to keep the version in sync _grin_

**Comment by tpmccallum** [comment]
Builds and runs:

```bash
$ spin build
Building component vault-variable-test with `cargo build --target wasm32-wasi --release`
    Finished release [optimized] target(s) in 0.06s
Finished building all Spin components
$ spin up --runtime-config-file runtime_config.toml
Logging component stdio to ".spin/logs/"

Serving http://127.0.0.1:3000
Available Routes:
  vault-variable-test: http://127.0.0.1:3000 (wildcard)
```

Making the request does make an error, though:

```bash
thread '<unnamed>' panicked at 'could not get variable: Error::Undefined("no variable for \"vault-variable-test\".\"test_password\"")', src/lib.rs:11:52
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2024-01-24T03:03:17.158681Z ERROR spin_trigger_http: Error processing request: guest invocation failed

Caused by:
    0: error while executing at wasm backtrace:
           0: 0x15676 - <unknown>!__rust_start_panic
           1: 0x15431 - <unknown>!rust_panic
           2: 0x1540c - <unknown>!std::panicking::rust_panic_with_hook::h59179097e3fe00d4
           3: 0x14772 - <unknown>!std::panicking::begin_panic_handler::{{closure}}::h685c1054195e4d4b
           4: 0x1469f - <unknown>!std::sys_common::backtrace::__rust_end_short_backtrace::hd312084468c4357c
           5: 0x14d22 - <unknown>!rust_begin_unwind
           6: 0x22cb6 - <unknown>!core::panicking::panic_fmt::h826facefad16f5a7
           7: 0x24fc8 - <unknown>!core::result::unwrap_failed::hf48c8310675f070e
           8: 0x560f - <unknown>!vault_variable_test::handle_vault_variable_test::hd63e5b2c7245f8f7
           9: 0x3bb6 - <unknown>!spin_sdk::http::executor::run::hd6b683a35367e806
          10: 0x5901 - <unknown>!wasi:http/incoming-handler@0.2.0-rc-2023-10-18#handle
       note: using the `WASMTIME_BACKTRACE_DETAILS=1` environment variable may show more debugging information
    1: wasm trap: wasm `unreachable` instruction executed    

```

File: examples/vault-variable-test/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "vault-variable-test"
+authors = ["tpmccallum <tim.mccallum@fermyon.com>"]
```
Comment (itowlson) [inline examples/vault-variable-test/Cargo.toml:3]: I'd suggest removing this from examples in the Spin repo (although I don't have strong feelings, just glancing at a couple of other random examples)

File: examples/vault-variable-test/spin.toml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+spin_manifest_version = 2
```
Comment (itowlson) [inline examples/vault-variable-test/spin.toml:1]: I might be missing something but the manifest doesn't seem to define the variables?

File: examples/vault-variable-test/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+description = "A simple application that returns query values from variable providers"
```
Comment (itowlson) [inline examples/vault-variable-test/Cargo.toml:4]: I don't think the description quite matches what the app does?

File: examples/vault-variable-test/spin.toml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+[application]
+name = "vault-variable-test"
+version = "0.1.0"
+authors = ["Fermyon Engineering <engineering@fermyon.com>"]
+description = "A simple application that returns query values from variable providers"
```
Comment (itowlson) [inline examples/vault-variable-test/spin.toml:7]: Same comment re description

File: examples/vault-variable-test/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+use spin_sdk::{
+    http::{IntoResponse, Request, Response},
+    http_component,
+    variables,
+};
+
+#[http_component]
+fn handle_vault_variable_test(req: Request) -> Result<impl IntoResponse> {
+    let password = std::str::from_utf8(req.body().as_ref()).unwrap();
+    let expected = variables::get("test_password").expect("could not get variable");
```
Comment (itowlson) [inline examples/vault-variable-test/src/lib.rs:11]: Worth commenting this line to draw attention to it?

File: examples/vault-variable-test/runtime_config.toml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+[[config_provider]]
```
Comment (itowlson) [inline examples/vault-variable-test/runtime_config.toml:1]: Is it worth adding a comment to this file?  Or talk about it in a README?  Not sure, just a vague wonder...

**Review by itowlson — CHANGES_REQUESTED** [review]
The content is fine, just a few possible issues (the key one being the variables in `spin.toml`), sorry

File: examples/vault-variable-test/README.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+5. Run the application:
+
+```bash
+spin up --runtime-config-file runtime_config.toml
+```
+
+6. Test the application:
+
+```bash
+curl -i http://127.0.0.1:3000
```
Comment (itowlson) [inline examples/vault-variable-test/README.md:37]: needs moar request body

```suggestion
curl --data guess http://127.0.0.1:3000
```

Comment (itowlson) [reply examples/vault-variable-test/README.md:37]: oh this might need a docs update too

**Review by itowlson — APPROVED** [review]
there is one change needed but otherwise all good - thanks!

File: examples/vault-variable-test/README.md (hunk 8 above)
Comment (tpmccallum) [inline examples/vault-variable-test/README.md:37]:
```suggestion
$ curl localhost:3000 --data "test_password"
{"authentication": "accepted"}
```

**Review by itowlson — APPROVED** [review]

File: examples/vault-variable-test/README.md (hunk 8 above)
Comment (tpmccallum) [reply examples/vault-variable-test/README.md]: Great catch @itowlson 
Updated in last commit of the PR, thanks again !!!

File: examples/vault-variable-test/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,23 @@
+use anyhow::Result;
+use spin_sdk::{
+    http::{IntoResponse, Request, Response},
+    http_component,
+    variables,
+};
+
+#[http_component]
+fn handle_vault_variable_test(req: Request) -> Result<impl IntoResponse> {
+    let attempt = std::str::from_utf8(req.body().as_ref()).unwrap();
```
Comment (rylev) [inline examples/vault-variable-test/src/lib.rs:10]:
```suggestion
    let attempt = std::str::from_utf8(req.body()).unwrap();
```

File: examples/vault-variable-test/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    let expected = variables::get("password").expect("could not get variable");
```
Comment (rylev) [inline examples/vault-variable-test/src/lib.rs:11]: Nit: I don't think we should be encouraging code that can panic.
```suggestion
    let expected = variables::get("password").context("could not get variable")?;
```

File: examples/vault-variable-test/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,23 @@
+use anyhow::Result;
```
Comment (rylev) [inline examples/vault-variable-test/src/lib.rs:1]: This is needed to make a suggestion below work
```suggestion
use anyhow::{Result, Context};
```

**Review by rylev — APPROVED** [review]

**Pull request merged**